addons/controllers/calico/calicoconfig_controller.go

```diff
@@ -46,3 +88,181 @@ func (r *CalicoConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+	if err := r.CreateDataValuesSecret(ctx, cluster, calicoConfig, secretNamespacedName); err != nil {
+		return err
+	}
+
+	// add the reference to the data values secret into CalicoConfig Status field
+	calicoConfig.Status.SecretRef = &cniv1alpha1.DataValuesSecretRef{
+		Name:      secretNamespacedName.Name,
+		Namespace: secretNamespacedName.Namespace,
+	}
+
+	return nil
+}
+
+func (r *CalicoConfigReconciler) CreateDataValuesSecret(ctx context.Context,
+	cluster *clusterapiv1beta1.Cluster,
+	calicoConfig *cniv1alpha1.CalicoConfig,
+	secretNamespacedName types.NamespacedName) error {
+
+	// prepare the data values secret for CalicoConfig
+	calicoDataValuesSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretNamespacedName.Name,
+			Namespace: secretNamespacedName.Namespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+	}
+
+	calicoDataValuesSecretMutateFn := func() error {
+		calicoDataValuesSecret.Data = make(map[string][]byte)
+		clusterNetwork := cluster.Spec.ClusterNetwork
+		if clusterNetwork == nil {
+			err := errors.New("cluster.Spec.ClusterNetwork is not set")
+			r.Log.Error(err, fmt.Sprintf("Could not find ClusterNetwork in cluster '%s' Spec", cluster.Name))
+			return err
+		}
+		clusterCIDR := clusterNetwork.Pods
+		if clusterCIDR == nil {
+			clusterCIDR = clusterNetwork.Services
+		}
+		if clusterCIDR == nil {
+			err := errors.New("cluster.Spec.ClusterNetwork.Pods and cluster.Spec.ClusterNetwork.Services are not set")
+			r.Log.Error(err, fmt.Sprintf("Could not find ClusterNetwork.Pods and ClusterNetwork.Services in cluster '%s' Spec", cluster.Name))
+			return err
+		}
+		m := structs.Map(calicoConfig.Spec)
+		calicoEntry := m["Calico"]
+		configEntry := calicoEntry.(map[string]interface{})["Config"]
+		configEntry.(map[string]interface{})["ClusterCIDR"] = clusterCIDR
+		yamlBytes, err := yaml.Marshal(m)
+		if err != nil {
+			r.Log.Error(err, "Error marshalling CalicoConfig to Yaml")
+			return err
+		}
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
+		calicoDataValuesSecret.Data[constants.DataValueFileName] = dataValueBytes
+		return nil
+	}
+
+	// create the data values secret for CalicoConfig
+	result, err := controllerutil.CreateOrPatch(ctx, r.Client, calicoDataValuesSecret, calicoDataValuesSecretMutateFn)
+	if err != nil {
+		r.Log.Error(err, "Error creating or patching CalicoConfig addon data values secret")
+		return err
+	}
+	r.Log.Info(fmt.Sprintf("Resource '%s' data values secret '%s'", constants.CalicoAddonName, result))
+
+	return nil
+}
+
+// ReconcileCalicoConfigDelete reconciles CalicoConfig deletion by deleting data values secret
+func (r *CalicoConfigReconciler) ReconcileCalicoConfigDelete(
+	ctx context.Context,
+	calicoConfig *cniv1alpha1.CalicoConfig,
+	patchCRD *bool) (retErr error) {
+
+	// remove the reference to the data values secret from CalicoConfig Status field
+	calicoConfig.Status.SecretRef = &cniv1alpha1.DataValuesSecretRef{}
```

Comment:
suggestion: `calicoConfig` here is the original pointer assigned by `r.Client.Get`. Making changes directly to the original pointer is risky. The caller of this function should make a deep copy at the very beginning and pass the copy down to `ReconcileCalicoConfigDelete`

Reply:
Done